File: java/Deepspace/src/deepspace/Weapon.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 *
 * @author alvaro
 */
public class Weapon implements CombatElement{   //quito el public?
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon (String na, WeaponType ty, int us){
        name = na;
        type = ty;
        uses = us;
    }
    
    Weapon (Weapon copia){
        this(copia.name, copia.type, copia.uses);
        /*name = copia.name;
        type = copia.type;
        uses = copia.uses;*/
    }
    
    public WeaponType getType(){
        return type;
    }
    
    @Override
    public int getUses(){
        return uses;
    }
    
    public float power(){
        return type.getPower();
    }
    
    @Override
    public float useIt(){
        if(uses > 0){
            uses--;
            return power();
        }else
            return 1.0f;
    }
    
    WeaponToUI getUIversion(){
        return new WeaponToUI(this);
    }
    
    //EXTRA: PARA VER FACILMENTE EL CONTENIDO DE UNA INSTANCIA DE WEAPON:
    @Override
    public String toString(){
        String result = "name: "  + this.name + "\n";
        result  = result + "    Type: " + getType() + "\n";
        result  = result + "    Uses: " + getUses() + "\n";
        result  = result + "    Power: " + power() + "\n";
        
        return result;
    }
}

File: ruby/Weapon.rb
#encoding:utf-8

require_relative 'WeaponToUI'

module Deepspace
    class Weapon
        def initialize(na, ty, us) #string, weapontype, int
            @name = na
            @type = ty
            @uses = us
        end

        ####################################
        #no estoy seguro de si es asi
        def type
            @type
        end
        ######################################

        #no estoy seguro de si hace falta: si, para el constructor de copia
        def name
            @name
        end

        def uses
            @uses
        end

        def power
            type.power
        end

        def self.newCopy(copia)
            Weapon.new(copia.name, copia.type, copia.uses)
        end

        def useIt
            if(uses > 0)
                @uses -= 1
                power
            else
                return 1.0
            end
        end

        def to_s
            if (type == Deepspace::WeaponType::LASER)
                t = "LASER"
            elsif (type == Deepspace::WeaponType::MISSILE)
                t = "MISSILE"
            else
                t = "PLASMA"
            end
            return "Name: #{@name}, Type: " + t + ", Uses: #{@uses}"
        end

        def getUIversion
            return WeaponToUI.new(self)
        end
    end
end